Item 9584: ManageSampleStatusesPanel for sample statuses CRUD within a given container

cnathe opened the pull request:
#### Rationale
Recent stories have added a Sample Status column and support for status change checks for permitted operations. This PR adds the UI component to manage the sample statuses available for a given container. It will be shown in LKS, LKSM, and LKB. In addition, this story adds in a SampleStatusService registered from the SM module so that by default the sample status support is not "enabled" unless the SM module is included in the build and registers the service implementation to support it.

#### Related Pull Requests
* https://github.com/LabKey/labkey-ui-components/pull/653
* https://github.com/LabKey/platform/pull/2698
* https://github.com/LabKey/sampleManagement/pull/726
* https://github.com/LabKey/biologics/pull/1026
* https://github.com/LabKey/inventory/pull/323

#### Changes
* ManageSampleStatusesPanel for sample statuses CRUD operations
* Update APIWrapper to take mockFn as a param instead of adding jest dependency directly
* getSampleStatuses action to call API and return SampleState array
* NameIdSettings component update to support optional titleCls prop

cnathe committed:
Item 9584: ManageSampleStatesPanel to display, update, add, remove sample statuses for a container
- TODO: save changes to the server for insert/update/delete

cnathe committed:
v2.84.0-fb-smManageStatuses.0

cnathe committed:
ManageSampleStatesPanel insert and update rows on save

cnathe committed:
v2.84.0-fb-smManageStatuses.1

cnathe committed:
ManageSampleStatesPanel delete rows with confirm modal
- and change some text from "states" to "statuses"

cnathe committed:
v2.84.0-fb-smManageStatuses.2

cnathe committed:
Reorder definition of components in ManageSampleStatusesPanel.tsx file

cnathe committed:
v2.84.0-fb-smManageStatuses.3

cnathe committed:
Use mockFn param to avoid jest depencency on package usage

cnathe committed:
linting

cnathe committed:
v2.84.0-fb-smManageStatuses.4

cnathe committed:
v2.84.0-fb-smManageStatuses.5

cnathe committed:
in-use -> in use

cnathe committed:
Merge branch 'develop' into fb_smManageStatuses

# Conflicts:
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
v2.85.0-fb-smManageStatuses.0

cnathe committed:
v2.85.0-fb-smManageStatuses.1

cnathe committed:
NameIdSettings.tsx component change to allow for titleCls optional prop

cnathe committed:
release notes update

cnathe committed:
remove unused ref prop

labkey-susanh commented on packages/components/src/internal/components/samples/APIWrapper.ts:
```diff
@@ -19,19 +21,26 @@ export interface SamplesAPIWrapper {
+        getSampleAliquotRows: mockFn(),
+        getSampleAssayResultViewConfigs: mockFn(),
+        getSampleSelectionLineageData: mockFn(),
+        getSampleStatuses: mockFn(),
```
If the same function is used for all of these, then doesn't that make testing whether the right method got called the right number of times impossible?

cnathe replied:
This `mockFn` is the default, but I believe in that case where you wanted to do that type of "did the right method get called the right number of times", then you would use `getTestAPIWrapper` and `getSamplesTestAPIWrapper` to provide your own mock for that specific function you want to track and then have the jest test ask for how many times that was called.

labkey-susanh replied:
Ah, OK. This is just to eliminate the dependency then.

labkey-susanh commented on packages/components/src/internal/components/samples/ManageSampleStatusesPanel.tsx:
```diff
@@ -0,0 +1,376 @@
+        setDirty(addNew);
+        setSaving(false);
+        setShowDeleteConfirm(false);
+        setError(undefined);
```
It's strange to me that these are set in this useEffect instead of just used in the constructor above. I guess this is a way to reinitialize things when `state` or `typeOptions` changes but should these values also be used in initialization above?

cnathe replied:
yeah, this is a little odd. this is more of a "reset state" after the save or delete action is complete. I'll move these to a function so the code is a bit more self documenting.

labkey-susanh commented on packages/components/src/internal/components/samples/ManageSampleStatusesPanel.tsx:
```diff
@@ -0,0 +1,376 @@
+                    setShowDeleteConfirm(false);
+                });
+        } else {
+            setShowDeleteConfirm(false);
```
Do you want to change the updatedState state variable?

cnathe replied:
no. I don't think so. this else block should really never be called since the delete button won't be shown for states that don't have a RowId value. so this was just a bit of a safe guard.

labkey-susanh replied:
Got it.  Just checking.

cnathe committed:
Move around some imports to prevent circular dependency issue in jest test

labkey-susanh commented on packages/components/src/internal/components/samples/ManageSampleStatusesPanel.tsx:
```diff
@@ -0,0 +1,376 @@
+                        </div>
+                    </FormGroup>
+                    <div>
+                        {!addNew && (
```
Once you click "Add", it seems there's no going back.  I think there should be a "Cancel" button shown if you're in "Add" mode since "Save" produces an error.

cnathe replied:
Cancel button added for "add new" case

labkey-susanh commented on packages/components/src/internal/components/samples/ManageSampleStatusesPanel.tsx:
```diff
@@ -0,0 +1,376 @@
+                    confirmButtonText="Yes, Delete"
+                    onCancel={onToggleDeleteConfirm}
+                    onConfirm={onDeleteConfirm}
+                    title="Permanently delete status?"
```
Title case for this perhaps?  (That's what I've settled on for the various other modals I've updated on my branch.)

cnathe replied:
Done

labkey-susanh commented on packages/components/src/internal/components/samples/actions.ts:
```diff
@@ -756,3 +757,47 @@ export function getSampleAssayResultViewConfigs(): Promise<SampleAssayResultView
     });
 }
+
+export class SampleState {
```
Should this be in the models file instead?

cnathe replied:
Yes. Moved.

labkey-susanh commented on packages/components/src/internal/components/samples/actions.ts:
```diff
@@ -756,3 +757,47 @@ export function getSampleAssayResultViewConfigs(): Promise<SampleAssayResultView
+export function getSampleStatuses(): Promise<SampleState[]> {
+    return new Promise((resolve, reject) => {
+        return Ajax.request({
+            url: buildURL('sampleManager', 'getSampleStatuses.api'),
```
We do have a constant for the 'sampleManager' controller somewhere, but no big deal.

cnathe replied:
Thanks for the reminder. Done.

cnathe committed:
Jest tests for ManageSampleStatusesPanel

cnathe committed:
Jest tests for SampleStatusesListItem

cnathe committed:
Jest tests for SampleStatusesListItem

cnathe committed:
Jest tests for SampleStatusDetail

cnathe committed:
Cr feedback
- move SampleState to models.ts file
- use SAMPLE_MANAGER_APP_PROPERTIES.controllerName
- ManageSampleStatusesPanel.tsx add Cancel button for add new case

cnathe committed:
Jest test update for ManageSampleStatusesPanel add new Cancel button

cnathe committed:
v2.85.0-fb-smManageStatuses.2

cnathe committed:
Trim status label before saving

cnathe committed:
v2.85.0-fb-smManageStatuses.3

cnathe committed:
Add initial dasbhoard chart config for sample counts by status

cnathe committed:
v2.85.0-fb-smManageStatuses.4

labkey-susanh commented on packages/components/src/internal/components/chart/configs.ts:
```diff
@@ -47,4 +47,18 @@ export const CHART_GROUPS: Record<string, ChartConfig> = {
         queryName: 'SampleSetCounts',
         schemaName: SCHEMAS.EXP_TABLES.SCHEMA,
     },
+    SampleStatuses: {
```
Not sure where this goes, but the styling on the bars in this chart shows the pointer as the cursor suggesting they are clickable.  Anything to be done about that?

cnathe replied:
Yes, fixed so that only the actually clickable boxes in the charts have that styling.

cnathe committed:
Update a few webpack README links to use experiment module instead of study

cnathe committed:
Update a few webpack README links

cnathe committed:
More minor webpack README updates

cnathe committed:
More minor webpack README updates

cnathe committed:
BarChartViewer updates to not show chart group selection when only one is present in group

cnathe committed:
yarn run lint-branch-fix

cnathe committed:
update release notes date and version

cnathe committed:
typo

cnathe committed:
v2.86.0
